"panic: invalid union tag" when attempting union() on single record

Repro is with Zed commit 2d52400.

Consider this test data, which I think is syntactically correct ZSON since it parses ok on its own.

```
$ cat data.zson 
{num: 123 (int32) ((int32,float64))}
{num: 456.0 (float32) ((int32,float32))}
{num: 789.0 (float64) ((float32,float64))}

$ zq -version
Version: v1.2.0-29-g2d52400f

$ zq -Z data.zson 
{
    num: 123 (int32) ((int32,float64))
}
{
    num: 456. (float32) ((int32,float32))
}
{
    num: 789. ((float32,float64))
}
```

If I `union()` multiple values, it works fine.

```
$ cat data.zson | zq -Z 'union(num)' -
{
    union: |[
        123 (int32) ((int32,float64)),
        456. (float32) ((int32,float32)),
        789. ((float32,float64))
    ]|
}

$ head -2 data.zson | zq -Z 'union(num)' -
{
    union: |[
        123 (int32) ((int32,float64)),
        456. (float32) ((int32,float32))
    ]|
}
```

But for some reason if I target just a single record, now I get a stack dump.

```
$ head -1 data.zson | zq -Z 'union(num)' -
panic: invalid union tag

goroutine 1 [running]:
github.com/brimdata/zed.(*TypeUnion).Untag(0xc0003f4d20, {0xc00037528a?, 0xc00019f4f8?, 0x100e807?})
	/Users/phil/work/zed/complex.go:387 +0x125
github.com/brimdata/zed/zson.(*elemHelper).add(0xc00019f5e8, {0xc00037528a?, 0x0?, 0x0?})
	/Users/phil/work/zed/zson/formatter.go:514 +0x48
github.com/brimdata/zed/zson.(*Formatter).formatVector(0xc0003ac320, 0x4, {0x17aada0?, 0x10?}, {0x17aad38, 0x2}, {0x193dce0, 0xc0003f4d20}, 0xc00019f710, 0x0, ...)
	/Users/phil/work/zed/zson/formatter.go:476 +0x548
github.com/brimdata/zed/zson.(*Formatter).formatValue(0xc0003ac320, 0xc0003ac360?, {0x193dcb8?, 0xc00031b590?}, {0xc000375289, 0x7, 0x7}, 0x0, 0x0, 0x1)
	/Users/phil/work/zed/zson/formatter.go:195 +0x756
github.com/brimdata/zed/zson.(*Formatter).formatRecord(0xc0003ac320, 0x0?, 0xc0003f5050?, {0xc000375288?, 0x8?, 0x8?}, 0x0?, 0x30?)
	/Users/phil/work/zed/zson/formatter.go:449 +0x835
github.com/brimdata/zed/zson.(*Formatter).formatValue(0xc0003ac320, 0xc0003f5050?, {0x193dc90?, 0xc0003f5050?}, {0xc000375288, 0x8, 0x8}, 0x0, 0x0, 0x0)
	/Users/phil/work/zed/zson/formatter.go:191 +0x6a5
github.com/brimdata/zed/zson.(*Formatter).formatValueAndDecorate(0xc0003ac320, {0x193dc90?, 0xc0003f5050?}, {0xc000375288, 0x8, 0x8})
	/Users/phil/work/zed/zson/formatter.go:162 +0x152
github.com/brimdata/zed/zson.(*Formatter).FormatRecord(0xc0003ac320, 0xc0003f50b0)
	/Users/phil/work/zed/zson/formatter.go:78 +0xaa
github.com/brimdata/zed/zio/zsonio.(*Writer).Write(0xc00031b1b8, 0xc00019fb00?)
	/Users/phil/work/zed/zio/zsonio/writer.go:33 +0x29
github.com/brimdata/zed/zbuf.WriteBatch({0x888a658, 0xc00031b1b8}, {0x1942248?, 0xc0003f50e0?})
	/Users/phil/work/zed/zbuf/batch.go:37 +0x77
github.com/brimdata/zed/zbuf.CopyPuller({0x888a658, 0xc00031b1b8}, {0x1938600, 0xc00038c500})
	/Users/phil/work/zed/zbuf/batch.go:110 +0x7c
github.com/brimdata/zed/cli/zq.(*Command).Run(0xc00029a000, {0xc0000360a0, 0x2, 0x2})
	/Users/phil/work/zed/cli/zq/command.go:166 +0x798
github.com/brimdata/zed/pkg/charm.path.run({0xc0000140c0?, 0x1, 0x1}, {0xc0000360a0?, 0x2, 0x0?})
	/Users/phil/work/zed/pkg/charm/path.go:11 +0x8b
github.com/brimdata/zed/pkg/charm.(*Spec).ExecRoot(0xc0000061a0?, {0xc000036090, 0x3, 0x3})
	/Users/phil/work/zed/pkg/charm/charm.go:63 +0x4a
main.main()
	/Users/phil/work/zed/cmd/zq/main.go:11 +0x5d
```

--- a/runtime/expr/agg/union.go
+++ b/runtime/expr/agg/union.go
@@ -64,19 +64,20 @@ func (u *Union) Result(zctx *zed.Context) *zed.Value {
 	for typ := range u.types {
 		types = append(types, typ)
 	}
-	types = zed.UniqueTypes(types)
-	inner := types[0]
-	if len(types) > 1 {
-		inner = zctx.LookupTypeUnion(types)
-	}
+	var inner zed.Type
 	var b zcode.Builder
-	for typ, m := range u.types {
-		for v := range m {
-			if union, ok := zed.TypeUnder(inner).(*zed.TypeUnion); ok {
+	if len(types) > 1 {
+		union := zctx.LookupTypeUnion(types)
+		inner = union
+		for typ, m := range u.types {
+			for v := range m {
 				zed.BuildUnion(&b, union.TagOf(typ), []byte(v))
-			} else {
-				b.Append([]byte(v))
 			}
+		}
+	} else {
+		inner = types[0]
+		for v := range u.types[inner] {
+			b.Append([]byte(v))
 		}
 	}
 	return zed.NewValue(zctx.LookupTypeSet(inner), zed.NormalizeSet(b.Bytes()))

--- a/runtime/expr/agg/ztests/union-single-union-type.yaml
+++ b/runtime/expr/agg/ztests/union-single-union-type.yaml
@@ -1,0 +1,8 @@
+zed: union(this) | yield union
+
+input: |
+  1((int64,string))
+  1((int64,string))
+
+output: |
+  |[1]|(|[(int64,string)]|)

